### DISCUSSION TITLE
Self hosting questions.

### DISCUSSION THREAD

**Discussion by SeaPanda117** [ente-io/ente#2318] [General]
I have searched through the discussions for “self host” and “docker” and couldn’t find these answers, but I apologize if I missed them and asked again. 

I would love to self host this, but am new to self hosting. 

If I follow these instructions [https://github.com/ente-io/ente/blob/main/server/docs/docker.md](url) will the docker container update as Ente updates or will it always be the version that exists at the time the container is created? How can this be manipulated to point the storage of Ente to an SMB share that’s on a truenas or Unraid machine? Do I just need to create a VM, install Docker, and make sure the VM just always has enough storage available for however many photos I have? 

I know that these questions will seem basic, but they honestly are a complete mystery to a beginner. (Also feel free to recommend resources where I can learn more about docker in general!)

Also, thank you to Ente for open sourcing such a legit project (even if I don’t fully understand how to implement it yet 😅). My intent is to learn how to self host this while paying an annual 2TB plan to support the project as it continues! Nice to have a legit alternative to Google Photos!

**Comment by mnvr** [comment]
> will the docker container update as Ente updates or will it always be the version that exists at the time the container is created? 

the second - "it always be the version that exists at the time the container is created". To update to the latest, you can:

1. If you're using the pre-built docker images, then you can do a `docker compose stop` (to stop your cluster), `docker compose pull` (to fetch the latest image), then `docker compose up -d` again to restart it

2. If you're building from source, then you can do `docker compose stop` (to stop your cluster), `git pull` (to fetch the latest code), then `docker compose up --build -d` (to rebuild and restart it).

These steps will recreate your container, but since the data is kept it volumes (https://github.com/ente-io/ente/blob/main/server/compose.yaml#L86), when the new container starts with the new code from the updated image, it'll just use the existing data, so your data will be preserved across updates.

> How can this be manipulated to point the storage of Ente to an SMB share that’s on a truenas or Unraid machine? Do I just need to create a VM, install Docker, and make sure the VM just always has enough storage available for however many photos I have?
>
> I know that these questions will seem basic,

These are not basic questions though 🙂 and I can empathise with you trying to find a answers. But I can't answer them for you, since the approach depends on your risk-tolerance (of potential data loss) and convenience. 

If you just want to play around, then the second approach ("create a VM, install Docker, and make sure the VM just always has enough storage available for however many photos I have") is fine. 

For more serious use, you'll need to ensure that volumes that I mentioned - (a) the database (the postgres-data volume in the sample docker compose file), and (b) the object storage (this is where the encrypted blobs are stored, this is the minio-data volume in the sample docker compose file) are stored in a way that is not tied to the lifetime of your containers, and are also backed up somewhere.

A middle ground (not suggesting you do this necessarily, I'm just mentioning this so you're aware since this might be a possibility that you could help you get started) is to not rely on self hosted ente as the primary source, but instead do backups from your clients. That is:

1. Self host Ente, in whatever way
2. From one of your clients that connect to your self hosted Ente, do a continuous export of data. See https://help.ente.io/photos/migration/export/.

2 is anyways a good option anyways even if you have a more solid self hosted setup. This export will not be E2EE, but it is also not tied to Ente in any way, so even if your self hosted setup breaks, you should be able to reimport this into Ente (or wherever else you wish).

**Reply by SeaPanda117** [reply]
Thank you for the reply. After reading it I’ve been googling and YouTubing furiously until now. 🤣
If the data is stored in volumes, shouldn’t we be able to mount a share (regardless of type) inside the VM and then mount that as a volume? The problem is that I don’t know how to mount that volume. 

It’s quite frustrating to feel like I’m almost right there and understanding is just over the nest hill. Then there is the knowledge that there is another hill right beyond that one. 🤣

**Reply by mnvr** [reply]
> The problem is that I don’t know how to mount that volume.

You can use the `external` flag in your docker compose file next to the postgres and minio volumes to let Docker know that these already exist on your machine.

See:

* https://docs.docker.com/storage/volumes/
* https://docs.docker.com/compose/compose-file/07-volumes/#external

**Comment by PaWyss** [comment]
Just a question to "git pull" and "docker compose up --build -d" (sorry, I'm no git user ;) )

Actually I'm running Ente self-hosted with the pre-build Docker image (by the way, great work for the product and the possibility to self-host it!!) . This image is some weeks behind.

If I'm doing a "git pull" and use in my compose.yaml the following lines:

"
    build:
      context: .
      args:
        GIT_COMMIT: development-cluster
"
Which version do I have then? A untested developer branch? Or what should I use for my daily usage? Maybe the pre-build image from you? :)

BR

**Reply by maffinca69** [reply]
@mnvr In my opinion, it would be cool to have a CI/CD to build docker images from the main branch (or any other stable one)

It seems that the assembly is now done manually, given the comment above

**Reply by mnvr** [reply]
Not sure if I'm following you. The Docker images are already built using a GitHub workflow from the main branch. We just don't do it on every merge.

If you want to do your own CI/CD, you can use the GitHub workflows that we use:
- This is the one we use for creating internal Docker images - https://github.com/ente-io/ente/blob/main/.github/workflows/server-release.yml
- This is the one we use for publishing Docker images to ghcr.io - https://github.com/ente-io/ente/blob/main/.github/workflows/server-publish.yml

**Reply by maffinca69** [reply]
@mnvr why your not automatic build releases from main branch?  How stability main branch? Safely use custom ci/cd to main branch?

**Reply by mnvr** [reply]
Automatic releases for every commit to main would be noisy, which is why we don't do it. In terms of stability, there are no guarantees for main, but we do make our best effort to ensure main is stable. Even with the published docker images, there are no guarantees, just that we usually publish them after they've been running our production servers which vets them a bit more.
